src/parser/error.rs

```diff
@@ -0,0 +1,190 @@
+        input: I,
+        f: Fun,
+    ) -> IResult<I, A, Self> {
+        match res {
+            Ok(a) => Ok((input, a)),
+            Err(e) => Err(Err::Error(ParseError::new(input, f(e)))),
+        }
+    }
+}
+
+impl<'a, F: LurkField> fmt::Display for ParseError<Span<'a>, F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut res = String::new();
+
+        writeln!(
+            &mut res,
+            "at line {}:{}",
+            self.input.location_line(),
+            self.input.get_column()
+        )?;
+        let line = String::from_utf8_lossy(self.input.get_line_beginning());
+
+        writeln!(&mut res, "{} | {}", self.input.location_line(), line)?;
+
+        let cols = format!("{} | ", self.input.location_line()).len() + self.input.get_column();
+        for _ in 0..(cols - 1) {
+            write!(&mut res, " ")?;
+        }
+        writeln!(&mut res, "^")?;
+
+        if let Some(exp) = self.expected {
+            writeln!(&mut res, "Expected {}", exp)?;
+        }
+
+        let mut errs = self.errors.iter().filter(|x| !x.is_nom_err()).peekable();
+        match errs.peek() {
+            // TODO: Nom verbose mode
+            None => writeln!(&mut res, "Internal parser error")?,
+            Some(_) => {
+                writeln!(&mut res, "Reported errors:")?;
+                for kind in errs {
+                    writeln!(&mut res, "- {}", kind)?;
+                }
+            }
+        }
+
+        write!(f, "{}", res)
+    }
+}
+
+impl<I: AsBytes, F: LurkField> nom::error::ParseError<I> for ParseError<I, F>
+where
+    I: InputLength,
+    I: Clone,
+{
+    fn from_error_kind(input: I, kind: ErrorKind) -> Self {
+        ParseError::new(input, ParseErrorKind::Nom(kind))
+    }
+
+    fn append(input: I, kind: ErrorKind, mut other: Self) -> Self {
+        match input.input_len().cmp(&other.input.input_len()) {
+            Ordering::Less => ParseError::new(input, ParseErrorKind::Nom(kind)),
+            Ordering::Equal => {
+                other.errors.push(ParseErrorKind::Nom(kind));
+                other
+            }
+            Ordering::Greater => other,
+        }
+    }
+
+    fn or(self, mut other: Self) -> Self {
+        match self.input.input_len().cmp(&other.input.input_len()) {
+            Ordering::Less => self,
+            Ordering::Equal => {
+                for x in self.errors {
+                    other.errors.push(x);
+                }
+                other
+            }
+            Ordering::Greater => other,
+        }
+    }
+}
+
+impl<I: AsBytes, F: LurkField> nom::error::ContextError<I> for ParseError<I, F>
+where
+    I: InputLength,
+    I: Clone,
+{
+    fn add_context(input: I, ctx: &'static str, other: Self) -> Self {
+        match input.input_len().cmp(&other.input.input_len()) {
+            Ordering::Less => ParseError {
+                input,
+                expected: Some(ctx),
+                errors: vec![],
```

Comment:
Why is this empty rather than self.errors?

Reply:
Because there's no `self` parameter here. `nom::error::ContextError` is really not actually very powerful https://docs.rs/nom/latest/nom/error/trait.ContextError.html, I'm looking at swapping in https://docs.rs/nom-supreme/latest/nom_supreme/context/trait.ContextError.html in the future which is a little better